Add T6615

tylermenezes opened the pull request:
Initial support for the T6615 CO2 sensor. Only reading CO2 is supported but documented commands for controlling ABC and elevation calibration are included for future work.

## Description:


**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#692

## Checklist:
  - [X] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

tylermenezes committed:
Add T6615

Initial support for the T6615 CO2 sensor. Only reading CO2 is supported
but documented commands for controlling ABC and elevation calibration
are included for future work.

tylermenezes committed:
Code quality cleanups

tylermenezes committed:
Remove unneeded python imports

pkuehne commented on esphome/components/t6615/t6615.cpp:
```diff
@@ -0,0 +1,79 @@
+
+void T6615Component::update() {
+  // CO2 sensor
+  if (this->co2_sensor_ != nullptr) {
```
Rather than wrapping the entire function in an `if` statement you could use an early return and reduce the amount of indentation needed!
```suggestion
  if (this->co2_sensor_ == nullptr) {
      return;
  }
```

tylermenezes replied:
The main reason I didn't make this an early-return was in case, in the future, it should publish any of the other values available to the sensor (elevation, sensor state, etc). I can change this and it could be changed back in the future though.

pkuehne replied:
It was only a suggestion 👍 If there are more values, I would put them each into a dedicated function and then call them from the `update()` function. Then each function would have it's own early return and it all still works.

pkuehne commented on esphome/components/t6615/t6615.cpp:
```diff
@@ -0,0 +1,79 @@
+  uint8_t response[T6615_RESPONSE_BUFFER_LENGTH];
+  uint8_t length = this->t6615_write_command_(sizeof(T6615_COMMAND_GET_PPM), T6615_COMMAND_GET_PPM, response);
+  if (length == 0) {
+    return 0;
```
This doesn't make a difference between failing to read a value and reading a value of 0. Is it definitely impossible to read a CO2 value of 0ppm?

tylermenezes replied:
It's probably not impossible, but I think the sensor uses 0 for other failure states. Is there a recommended way to do this in C++? I'm not very experienced with C++ coding practices.

pkuehne replied:
Simplest would probably be to use negative numbers for error states, like -1 for failed to write command, -2 for invalid size, etc. Otherwise, you could take the return value as a reference/pointer and return a boolean for success/failure.
e.g.
```cpp
bool T6615Component::get_ppm_(uint16_t& value)
```

pkuehne commented on esphome/components/t6615/t6615.cpp:
```diff
@@ -0,0 +1,79 @@
+  // Read header
+  uint8_t header[3];
+  this->read_array(header, 3);
+  if (header[0] != 0xFF || header[1] != 0xFA) {
```
Could these values be made named constants? That will make it clearer why the `header` shouldn't have those values.

tylermenezes committed:
Code quality updates for T6615

- Switch int get_ppm for void publish_ppm method to simplify code
- Document magic number and addresses as constants

tylermenezes committed:
Clang format cleanup

OttoWinter commented on esphome/components/t6615/t6615.cpp:
```diff
@@ -0,0 +1,77 @@
+  this->write_byte(len);
+  this->write_array(command, len);
+  this->flush();
+  delay(40);
```
This delay is too long to be run in the main loop.

Please use `set_timeout` instead here.

jesserockz committed:
Merge branch 'dev' into pr/tylermenezes/1170

jesserockz committed:
Future proof this a little bit
Add test
Add CODEOWNER

jesserockz reviewed (approved):
@tylermenezes I hope you don't mind. I was going through older PRs that could be updated and merged and made a few changes here. Please comment and confirm this works for you and I'll merge it in.

This should also make it easier to add the other functions later.

jesserockz committed:
Add const to TAG

tylermenezes commented:
Cool thank you! I never had the time to refractor this as people asked.

OttoWinter reviewed (approved):
LGTM, thanks! :tada:
